File: Python file/menstrualcycleapp.py
from datetime import datetime, timedelta 

cycle_length = 0
bleeding_length = 0

def date_validation(date):
	date = date.replace(",", " ").replace("-", " ").replace("/", " ").replace("_", " ")
	date_divid = date.split()
	if len(date_divid) < 3:
		raise ValueError("Invalid date format")
	for count in range(3):
		if not date_divid[count].isdigit():
			raise ValueError("Invalid date format: Non-numeric values detected")
	year, month, day = int(date_divid[0]), int(date_divid[1]), int(date_divid[2])
	if year != 2025:
		raise ValueError("Invalid year")	
	elif month > 12 or month < 1:
		raise ValueError("Invalid month")
	elif day > 31 or day < 1:
		raise ValueError("Invalid day")
	else:
		date = datetime(year, month, day)
	return date


def first_day(date):
	start_date = date_validation(date)
	return start_date
	
def calculate_cycle_length(start_date):
	if start_date > 35 or start_date < 21:
		return "invalid input"
	global cycle_length
	cycle_length = start_date	
	return cycle_length

def calculate_bleeding_length(number):
	global bleeding_length
	if number > 7 or number < 1:
		return "invalid input"
	bleeding_length = number
	return bleeding_length

def calculate_next_day(first_day, calculate_cycle_length):
	next_period_date = start_date + timedelta(days= cycle_length)
	return next_period_date

menu = '''
===========================================
	Welcome To Menstrual Cycle Calculator
	
	Press:
	1. Check your Menstrual Cycle
	0. Exit
===========================================
	'''
	
exit = True
while exit:
	print(menu)
	menu_input = input()
	match menu_input:
		case '1':
			inner_menu = '''
===========================================
	Menstrual Cycle Result Menu
	
	Press:
	1. View all the Result
	2. View Your Next Period Date
	3. View Your Ovulation Date
	4. View Your Safe Day
	5. View Your Fertile Window
	0. <= Back
	
===========================================
			
					'''
			exit1 = True
			while exit1:
				print("Enter Last Date Period Started (eg. 2025-01-05) : ")
				start_date = first_day(input())
				print("Enter Cycle length (21 - 35):")
				calculate_cycle_length(int(input()))
				print("Enter bleeding length (1 - 7):")
				calculate_bleeding_length(int(input()))
				calculate_next_day(first_day, calculate_cycle_length)
				ovulation_day = calculate_next_day(first_day, calculate_cycle_length) - timedelta(days=14)
				fertile_window_start = ovulation_day - timedelta(days=5)
				fertile_window_end = ovulation_day + timedelta(days=1)
				first_safe_day_start = start_date + timedelta(days= bleeding_length + 1)
				first_safe_day_end = fertile_window_start - timedelta(days=1)
				second_safe_day_start = fertile_window_end + timedelta(days=1)
				second_safe_day_end = calculate_next_day(first_day, calculate_cycle_length) - timedelta(days=1)

				
				print(inner_menu)
				inner_menu_input = input()
				match inner_menu_input:
					case '1':
						print("Your next period date: ", calculate_next_day(first_day, calculate_cycle_length))
						print("Your estimated ovulation date:", ovulation_day)
						print("Your estimated First Safe date:", first_safe_day_start, "to", first_safe_day_end)
						print("Your estimated Second Safe date:", second_safe_day_start, "to", second_safe_day_end)
						print("Your fertile window date starts: ", fertile_window_start)
						print("Your fertile window date ends: ", fertile_window_end)
						break
					case '2':
						print("Your next period date: ", calculate_next_day(first_day, calculate_cycle_length))
						break
					case '3':
						print("Your estimated ovulation date:", ovulation_day)
						break
					case '4':
						print("Your estimated First Safe date:", first_safe_day_start, "to", first_safe_day_end)
						print("Your estimated Second Safe date:", second_safe_day_start, "to", second_safe_day_end)
						break
					case '5':
						print("Your fertile window date starts: ", fertile_window_start)
						print("Your fertile window date ends: ", fertile_window_end)
						break
					case '0':
						exit1 = False
						
		case '0':
			exit = False

# handle your validation properly,
# what happens if i dont enter date in your format
# your test are not testing functionalities, just validation

File: Java file/MenstrualCycleApp.java
import java.util.Scanner;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;

public class MenstrualCycleApp {
    public static void main(String[] args) {
        Scanner prompt = new Scanner(System.in);

        boolean exit = true;
        while (exit) {
            String menu = """
===========================================
    Welcome To Menstrual Cycle Calculator

    Press:
    1. Check your Menstrual Cycle
    0. Exit
===========================================
            """;

            System.out.print(menu);
            String menu_input = prompt.nextLine();
            switch (menu_input) {
                case "0":
                    exit = false;
                    break;
                case "1":
                    System.out.print("Enter Last Date Period Started (YYYY-MM-DD): ");
                    LocalDate dateInput = MenstrualCycleFunction.date_validation(prompt.nextLine());
                    LocalDate start_date = dateInput;

                    System.out.print("Enter Cycle Length (1-35): ");
                    int cycle_length = Integer.parseInt(prompt.nextLine());

                    System.out.print("Enter Bleeding Length (1-7): ");
                    int bleeding_length = Integer.parseInt(prompt.nextLine());

                    LocalDate next_period_date = start_date.plusDays(cycle_length);
                    LocalDate ovulation_day = next_period_date.minusDays(14);
                    LocalDate fertile_window_start = ovulation_day.minusDays(5);
                    LocalDate fertile_window_end = ovulation_day.plusDays(1);
                    LocalDate first_safe_day_start = start_date.plusDays(bleeding_length + 1);
                    LocalDate first_safe_day_end = fertile_window_start.minusDays(1);
                    LocalDate second_safe_day_start = fertile_window_end.plusDays(1);
                    LocalDate second_safe_day_end = next_period_date.minusDays(1);

                    DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");

                    boolean exit1 = true;
                    while (exit1) {
                        String inner_menu = """
===========================================
    Menstrual Cycle Result Menu

    Press:
    1. View all the Results
    2. View Your Next Period Date
    3. View Your Ovulation Date
    4. View Your Safe Day
    5. View Your Fertile Window
    0. <= Back
===========================================
                        """;

                        System.out.print(inner_menu);
                        String inner_menu_input = prompt.nextLine();

                        switch (inner_menu_input) {
                            case "0":
                                exit1 = false;
                                break;
                            case "1":
                                boolean exit2 = true;
                                while (exit2) {
                                    System.out.println("Your next period date: " + next_period_date.format(formatter));
                                    System.out.println("Your estimated ovulation date: " + ovulation_day.format(formatter));
                                    System.out.println("Your First Estimated Safe Day before fertile window: " +
                         first_safe_day_start.format(formatter) + " to " + first_safe_day_end.format(formatter));
                                    System.out.println("Your Second Estimated Safe Day after fertile window: " +
                         second_safe_day_start.format(formatter) + " to " + second_safe_day_end.format(formatter));
                                    System.out.println("Your fertile window starts: " + fertile_window_start.format(formatter));
                                    System.out.println("Your fertile window ends: " + fertile_window_end.format(formatter));
                                    System.out.print("""
=============
Press:
0. <= Back
=============
""");
                                    String back_input = prompt.nextLine();
                                    if (back_input.equals("0")) exit2 = false;
                                    else System.out.println("""
*************
Invalid input
*************
""");
                                }
                                break;
                            case "2":
                                boolean exit3 = true;
                                while (exit3) {
                                    System.out.println("Your next period date: " + next_period_date.format(formatter));
                                    System.out.print("""
=============
Press:
0. <= Back
=============
""");
                                    String back_input = prompt.nextLine();
                                    if (back_input.equals("0")) exit3 = false;
                                    else System.out.println("""
*************
Invalid input
*************
""");
                                }
                                break;
                            case "3":
                                boolean exit4 = true;
                                while (exit4) {
                                    System.out.println("Your estimated ovulation date: " + ovulation_day.format(formatter));
                                    System.out.print("""
=============
Press:
0. <= Back
=============
""");
                                    String back_input = prompt.nextLine();
                                    if (back_input.equals("0")) exit4 = false;
                                    else System.out.println("""
*************
Invalid input
*************
""");
                                }
                                break;
                            case "4":
                                boolean exit5 = true;
                                while (exit5) {
                                    System.out.println("Your First Estimated Safe Day before fertile window: " +
                         first_safe_day_start.format(formatter) + " to " + first_safe_day_end.format(formatter));
                                    System.out.println("Your Second Estimated Safe Day after fertile window: " +
                         second_safe_day_start.format(formatter) + " to " + second_safe_day_end.format(formatter));
                                    System.out.print("""
=============
Press:
0. <= Back
=============
""");
                                    String back_input = prompt.nextLine();
                                    if (back_input.equals("0")) exit5 = false;
                                    else System.out.println("""
*************
Invalid input
*************
""");
                                }
                                break;
                            case "5":
                                boolean exit6 = true;
                                while (exit6) {
                                    System.out.println("Your fertile window starts: " + fertile_window_start.format(formatter));
                                    System.out.println("Your fertile window ends: " + fertile_window_end.format(formatter));
                                    System.out.print("""
=============
Press:
0. <= Back
=============
""");
                                    String back_input = prompt.nextLine();
                                    if (back_input.equals("0")) exit6 = false;
                                    else System.out.println("""
*************
Invalid input
*************
""");
                                }
                                break;
                        }
                    }
                    break;
            }
        }
       
    }
}
